### ISSUE TITLE
Partial date precision formatting in queries produces incorrect join.

### ISSUE THREAD

**Issue by benanhalt** [specify/specify7#654]
![Screenshot_2020-04-23_16-45-29](https://user-images.githubusercontent.com/287254/80152814-f8394680-8581-11ea-8ef0-71efa59d496c.png)

The above query should return a single result. Instead it returns a result for every collectingevent in the database. N.B. it only occurs if the partial date field occurs in a joined table rather than the base table of the query.

**Comment by benanhalt** [comment]
The bad generated SQL include a full cross join between collectionobject and collectingevent:

```
[23/Apr/2020 16:44:21] [DEBUG] [specifyweb.stored_queries.execution:512] query: SELECT collectionobject.`CollectionObjectID` AS `collectionobject_CollectionObjectID`, CAST(collectionobject.`CatalogNumber` AS DECIMAL(65)) AS anon_1, CASE collectingevent.`StartDatePrecision` WHEN %s THEN date_format(collectingevent_1.`StartDate`, %s) WHEN %s THEN date_format(collectingevent_1.`StartDate`, %s) WHEN %s THEN date_format(collectingevent_1.`StartDate`, %s) END AS anon_2 
FROM collectingevent, collectionobject LEFT OUTER JOIN collectingevent AS collectingevent_1 ON collectingevent_1.`CollectingEventID` = collectionobject.`CollectingEventID` 
WHERE collectionobject.`CollectionID` = %s AND collectionobject.`CatalogNumber` = %s
```

**Comment by benanhalt** [comment]
I cherry-picked this fix into the specifycloud server.

**Comment by maxxxxxdlp** [comment]
Confirmed fixed in the same database on the `develop` branch

I found few more problems (on the `master` branch, before the fix)

This endless loop happens even if there is no data specified 
<img width="1440" alt="Screen Shot 2020-04-24 at 10 03 35 AM" src="https://user-images.githubusercontent.com/40512816/80227898-ee602380-8613-11ea-90da-3a2166b86781.png">

Also, if there is only year specified as 2020, results show today's date when querying Full Data
<img width="1440" alt="Screen Shot 2020-04-24 at 10 04 29 AM" src="https://user-images.githubusercontent.com/40512816/80227899-eef8ba00-8613-11ea-9651-228c62e8296d.png">

Though none of these were reproducible on the `develop` branch, so they were probably fixed in 311c44f
